projects/kit/src/lib/processors/normalize-date-preprocessor.ts

```diff
@@ -14,53 +14,45 @@ export function normalizeDatePreprocessor({
-            : dateTimeSeparator;
-        const possibleDates = data.split(separator);
-        const dates = data.includes(dateTimeSeparator)
-            ? [possibleDates[0] ?? '']
-            : possibleDates;
+        const dateTimeSegments = data.split(/\D/).filter(Boolean);
+        const templateSegments = dateModeTemplate.split(dateSegmentsSeparator);
+        const includesTime = data.includes(dateTimeSeparator);
 
-        if (
-            dates.every(
-                (date) =>
-                    date.trim().split(/\D/).filter(Boolean).length ===
-                    dateModeTemplate.split(dateSegmentsSeparator).length,
-            )
-        ) {
-            const newData = dates
-                .map((date) =>
-                    normalizeDateString(date, dateModeTemplate, dateSegmentsSeparator),
-                )
-                .join(rangeSeparator);
+        const dateSegments = dateTimeSegments.slice(
+            0,
+            includesTime ? templateSegments.length : Infinity,
+        );
 
-            return {
-                elementState,
-                data: `${newData}${
-                    data.includes(dateTimeSeparator)
-                        ? dateTimeSeparator + possibleDates[1] || ''
-                        : ''
-                }`,
-            };
+        const timeSegments = dateTimeSegments.slice(
+            includesTime ? templateSegments.length : Infinity,
+        );
+
+        if (!dateSegments.length || dateSegments.length % templateSegments.length !== 0) {
+            return {elementState, data};
         }
 
-        return {elementState, data};
-    };
-}
+        const dates = dateSegments.reduce<string[]>((dates, segment, index) => {
+            const template = templateSegments[index % templateSegments.length];
+            const dateIndex = Math.trunc(index / templateSegments.length);
+            const isLastDateSegment =
+                index % templateSegments.length === templateSegments.length - 1;
 
-function normalizeDateString(
-    dateString: string,
-    template: string,
-    separator: string,
-): string {
-    const dateSegments = dateString.split(/\D/).filter(Boolean);
-    const templateSegments = template.split(separator);
-    const normalizedSegments = dateSegments.map((segment, index) =>
-        index === templateSegments.length - 1
-            ? segment
-            : segment.padStart(templateSegments[index]?.length ?? 0, '0'),
-    );
+            if (!dates[dateIndex]) {
+                dates[dateIndex] = '';
+            }
 
-    return normalizedSegments.join(separator);
+            dates[dateIndex] += isLastDateSegment
+                ? segment
+                : `${segment.padStart(template?.length ?? 0, '0')}${dateSegmentsSeparator}`;
+
+            return dates;
+        }, []);
+
+        return {
+            elementState,
+            data: includesTime
+                ? `${dates[0]}${dateTimeSeparator}${timeSegments.join(':')}`
```

Comment:
It is not true for all time modes – time mode can contain dot too.

https://github.com/taiga-family/maskito/blob/20fce011c67173ab3155a7145ae4299405943917/projects/kit/src/lib/types/time-mode.ts#L10

Reply:
```suggestion
                ? `${dates[0]}${data.slice(dateTimeSeparator)}`
```

\+ delete no more required `timeSegments` array

Reply:
Add some unit tests for `DateTime` (with `MM:SS.MSS` mode) inside `normalize-date-preprocessor.spec.ts`, please

Reply:
It also means that we can delete `dateTimeSegments` variable

```ts
const dateSegments = data
    .split(/\D/)
    .filter(Boolean)
    .slice(0, includesTime ? templateSegments.length : Infinity);
```